### ISSUE TITLE
Build the standard library with different `cfg` settings

### ISSUE THREAD

**Issue by ehuss** [rust-lang/wg-cargo-std-aware#4]
This is a "use case" issue to try to clarify and discuss a use case, and how it will be addressed.

There has been some interest in being able to pass different `cfg` flags to the standard library to change it some way; either to change the implementation, or to disable sections of the library.

This is unlikely to be possible with Cargo's current features system. There are some problems with backwards-compatibility, and generally the level of commitment to stabilize a feature flag would likely be extremely high.

The [RFC 2663 proposal](https://github.com/rust-lang/rfcs/pull/2663) had a suggestion of `force-tiny-fmt` to use a different algorithm for `fmt`.

`std` already has a number of cargo features (panic, backtrace, [etc.](https://github.com/rust-lang/rust/blob/95b1fe560d2bd8472f250fb8cfd2168520a58405/src/libstd/Cargo.toml#L62-L86)).

It would be good to collect some possible ideas people will need, and how these can possibly be expressed from Cargo, and whether or not it is feasible to accomplish in a stable way.

This is partially related to #x.

**Comment by SimonSapin** [comment]
If we expose some control over this, I think it will be limited to `#[cfg(feature = "…")]` like in "normal" dependency crates, not arbitrary `#[cfg(…)]` like the issue title and description suggest.

**Comment by ehuss** [comment]
> If we expose some control over this, I think it will be limited to …

Yes, that is what I meant.  I think this originally started as a broader thought about portability issues (`cfg` being influenced by target settings and `rustc` flags), but I didn't update the title when I narrowed it down.

**Comment by ehuss** [comment]
An experimental flag has been added to control the features used: [docs](https://doc.rust-lang.org/nightly/cargo/reference/unstable.html#build-std-features), [PR](https://github.com/rust-lang/cargo/pull/8490)

**Comment by Aaron1011** [comment]
Use case: Opt-in support for having `RefCell` capture backtraces for outstanding borrows, and print them when `borrow()`/`borrow_mut()` panics. See https://rust-lang.zulipchat.com/#narrow/stream/131828-t-compiler/topic/Attaching.20backtraces.20to.20RefCell/near/226273614 for more details

**Comment by macpp** [comment]
I'm not sure if this is intended or accidental (i didn't see it mentioned in unstable docs or cargo test suite) but it is currently possible to specify features on libcore directly, like this : `-Z build-std=std -Z build-std-features=core/debug_refcell`. 
In the case of `debug_refcell` it is actually usefull, because this feature is currently not forwarded by libstd and libtest, so just using `-Z build-std-features=debug_refcell` will not work

**Comment by ehuss** [comment]
I'm going to close as this was pushed off into a [future possibility](https://rust-lang.github.io/rfcs/3875-build-std-explicit-dependencies.html#allow-enablingdisabling-features-with-build-std) in the RFC. There are no immediate plans to do this for now.
